### ISSUE TITLE
gamma on linux

### ISSUE THREAD

**Issue by djmannion** [psychopy/psychopy#1618]
After [this report on discourse](https://discourse.psychopy.org/t/setting-the-gamma-ramp-manually/3543), I realised that setting the gamma does not work on linux.

Having investigated a bit, I think the problem comes down to the LUT size.

On one machine ("[AMD/ATI] Turks PRO [Radeon HD 6570/7570/8550]"), the LUT size seems to be 1024 rather than 256. Adjusting the code in ``gamma.py`` and passing an appropriate ramp then allows the gamma to be set.

On another machine ("Intel Corporation HD Graphics 520 (rev 07)"), the LUT size was being returned as 0. Changing the xorg driver fixed the LUT size to 256 and gamma functionality was restored.

Moving forward, it seems that we should adjust ``psychopy.visual.gamma`` to create and set the ramps based on a variable ramp size (with a warning or bail if the ramp size is 0).

Here's a draft/mockup of a ramp size getter:

```python
def getGammaRampSize(pygletWindow):
    
    rampSize = ctypes.c_int()

    success = xf86vm.XF86VidModeGetGammaRampSize(
        pygletWindow._x_display,
        pygletWindow._x_screen_id,
        ctypes.byref(rampSize)
    )

    assert success, "XF86VidModeGetGammaRampSize failed"

    return rampSize.value
```

It would also be good to get a sense from other linux users about whether gamma is working for them, given the apparent driver variability.

**Comment by djmannion** [comment]
The ramp size is 2048 (!) on my nvidia machine ("NVIDIA Corporation GK107 [GeForce GTX 650]").

**Comment by peircej** [comment]
Mario Kleiner might know how we can interrogate/set the size of the LUT

**Comment by djmannion** [comment]
Would you be happy for me to put together a draft PR that adapts the ramp to the LUT size? Might be useful even if we do end up being able to set the LUT size.

**Comment by peircej** [comment]
Most definitely! I assume this is a key step in allowing the higher-than-8-bit framebuffers (but maybe not all of it). eg. I assume your 2048 LUT is indicating a 10-bit output (or at least the ability to hold one)

**Comment by djmannion** [comment]
Closed via #1626.

**Comment by djmannion** [comment]
It looks like the gamma functionality is also broken under pygame. A look at the source code suggests that there are a few hardcoded 256s in there - presumably it can't cope with > 8-bit LUTs.

Should we use the gamma approach we use for pyglet windows for pygame also?

**Comment by djmannion** [comment]
N.B. the current code uses ```self.winHandle.set_gamma```, but that fails with an AttributeError. I think it needs to call ```pygame.display.set_gamma```.

**Comment by peircej** [comment]
Yes, the comment I made on whether setGamma should remain backend-dependent made me think and, actually, I think it could be a single setGamma for all engines so that we don't care whether other backend-providers have done it "incorrectly" in their own implementation. Given that the gamma affects the whole screen (not just the window) I don't think it will ever *require* a specific backend, so I think this will be fine.

**Comment by peircej** [comment]
i.e. lets drop all reference to pygame's own solution

**Comment by djmannion** [comment]
Sounds good. But at the moment I'm not sure how to get the equivalent of ```screenID``` and ```xDisplay``` from pygame - I'll keep investigating.

**Comment by djmannion** [comment]
I haven't found a good cross-platform way to get the required information when using a pygame window.

How about if we go back to requiring each backend to handle its own gamma? It looks like glfw has its own gamma routines. That would mean that gamma support in pygame would be limited to 8-bit LUTs, but that seems reasonable given its deprecation.

**Comment by peircej** [comment]
Sounds good

**Comment by djmannion** [comment]
I've had a go at this in #1714.

**Comment by hoechenberger** [comment]
Closed via #1732.
